### PR TITLE
introduce pytest.Marked as holder for marked parameter values

### PR DESCRIPTION
**Pull request by RonnyPfannschmidt**
this pr introduces a actual real type to separate parameter values and marks

this enables having functions as parameters with marks

it currently mirrors the broken mark overriding of the nested mark extraction code in place

documentation and examples upcoming

### CODE REVIEW

**Comment by RonnyPfannschmidt** [comment]
@flub the current way is seriously broken - its impossible to pass a function for example
marks should never ever made it to be used that way

the code that enables it is a painful hack i want to see gone

pytest.mark.foo is something we apply to functions, not to values

its a different type of behavior, and it breaks horrible

**Comment by flub** [comment]
Ah, I see what you mean.  I must admit my first reaction on marking a parameterised value with skipped was hesitation as well but then practicality seemed to beat purity.

But if we want to deprecate that why not experiment with other APIs?  how about for example:
`@pytest.mark.parametrize('a', [1, 2, 3], markers=[None, pytest.mark.skipif, None])`?  I think i mainly feel that the `pytest.marked()` api you're proposing here is not really addressing all the problems with marking parameters.

**Comment by RonnyPfannschmidt** [comment]
well, you can have single and multiple marks per value, and not all values have marks
so adding yet another list is a pain (from my pov)

**Comment by flub** [comment]
Any other creative alternative?  How about `@pytest.mark.parametrize(Parameter('name', 'id', marks=[a, b]), ...)`?

I realise I'm just throwing kind of random things at this but if changing the API anyway it's a good idea to brainstorm about how to create the best API with the requirements we know now.

**Comment by The-Compiler** [comment]
I don't really like the idea of having another list either - `Parameter` sounds interesting though, especially because it provides an alternative to `ids` as well.

I think that'd be really useful, for example in this qutebrowser test where it's tricky to keep the `ids` lined up with the tests:

```python

    @pytest.mark.parametrize('code', [
        str,
        lambda e: e[None],
        lambda e: operator.setitem(e, None, None),
        lambda e: operator.delitem(e, None),
        lambda e: None in e,
        list,  # __iter__
        len,
        lambda e: e.has_frame(),
        lambda e: e.geometry(),
        lambda e: e.style_property('visibility', strategy='computed'),
        lambda e: e.text(),
        lambda e: e.set_text('foo'),
        lambda e: e.insert_text('foo'),
        lambda e: e.is_writable(),
        lambda e: e.is_content_editable(),
        lambda e: e.is_editable(),
        lambda e: e.is_text_input(),
        lambda e: e.remove_blank_target(),
        lambda e: e.outer_xml(),
        lambda e: e.tag_name(),
        lambda e: e.rect_on_view(),
        lambda e: e._is_visible(None),
    ], ids=['str', 'getitem', 'setitem', 'delitem', 'contains', 'iter', 'len',
            'frame', 'geometry', 'style_property', 'text', 'set_text',
            'insert_text', 'is_writable', 'is_content_editable', 'is_editable',
            'is_text_input', 'remove_blank_target', 'outer_xml', 'tag_name',
            'rect_on_view', 'is_visible'])
    def test_vanished(self, elem, code):
        """Make sure methods check if the element is vanished."""
        elem._elem.isNull.return_value = True
        elem._elem.tagName.return_value = 'span'
        with pytest.raises(webkitelem.IsNullError):
            code(elem)
```

**Comment by RonnyPfannschmidt** [comment]
its relatively easy to add a name to a marked value as well

``` 
parametrize('a', [pytest.marked_value(name='tricky', marks=[pytest.mark.xfail, pytest.mark.long_running], value=42)])

**Comment by RonnyPfannschmidt** [comment]
we chould also extend the marked_value mechanism to a general annotated value

then users can put test ids# names, marks and whatever they want in

**Comment by flub** [comment]
Another use-case I got reminded off via another issue is that you can have `ids` to be a callable currently I think?.  The "marked value/param object" approach might make that one tricky.

**Comment by The-Compiler** [comment]
> we chould also extend the marked_value mechanism to a general annotated value
> then users can put test ids# names, marks and whatever they want in

That sounds like a quite confusing API, which adds another overload for the word `mark`. I really prefer @flub's suggestion about having a `pytest.Parameter`.

> Another use-case I got reminded off via another issue is that you can have ids to be a callable currently I think?. The "marked value/param object" approach might make that one tricky.

Why? I'd guess the callable would just get the value (and not the `Parameter` object) like before? The only open question would be what happens if there's a `Parameter('foo', id='bar')` with `ids=...` (with either an iterable or a callable).

**Comment by RonnyPfannschmidt** [comment]
so can we get back to making this something directly usable - namely - giving values a set of marks

that part is solved, and the name should be discussed

i rather want to introduce a simple straightforward api now than try to get a more complete concept right (which usually fails for a few iterations)

**Comment by flub** [comment]
I'm not at all convinced that the idea of a "marked value" is a very simple and straightforward concept.  I would actually have no idea what it is in isolation let alone guess that it is a way of being able to apply a mark to one of the generated test items/functions.

I think as the various parametrisation mechanisms have grown they have got weirder and weirder and are now really not as straight forward as might be desired for people new to it.

**Comment by RonnyPfannschmidt** [comment]
basically a marked value is intended to bring markers and parametrization together

perhaps it should be called marked_parameter instead of marked value,

in any case, what we currently have is broken and cant be fixed without breaking things worse

**Comment by nicoddemus** [comment]
Sorry for joining so late in the discussion.

I empathize with the fact that @RonnyPfannschmidt wants to solve the problem of "how to mark a value" quickly, but I also see the merits of the points raised by @flub and @The-Compiler that since we are introducing a new player into the mix (a "marked value"), we might as well discuss if we don't want to improve the overall API. We might arrive at something that not only fix how to mark values but also fixes some other problems/warts.

I agree that the current parametrization API might be a little awkward specially when multiple parameters, `ids`, and markers are involved. I like the `Parameter` idea at first sight and would like to see if we can evolve it a bit.

Here an example from the docs, adding `ids` for extra juice:

```python
@pytest.mark.parametrize(("n", "expected"), [
    (1, 2),
    pytest.mark.xfail((1, 0)),
    pytest.mark.xfail(reason="some bug")((1, 3)),
    (2, 3),
    (3, 4),
    (4, 5),
    pytest.mark.skipif("sys.version_info >= (3,0)")((10, 11)),
], ids=['1-2', '1-0', '1-3', '2-3', '3-4', '4-5', '10-11'])
def test_increment(n, expected):
    assert n + 1 == expected
```

Is something like the following what @flub and @The-Compiler are thinking about?

```python
@pytest.mark.parametrize(("n", "expected"), [
    pytest.param(1, 2, id='1-2'),
    pytest.param(1, 0, marks=pytest.mark.xfail, id='1-0'),
    pytest.param(1, 3, marks=pytest.mark.xfail(reason="some bug"), id='1-3'),
    pytest.param(2, 3, id='2-3'),
    pytest.param(3, 4, id='3-4'),
    pytest.param(4, 5, id='4-5'),
    pytest.param(10, 11, id='10-11', marks=pytest.mark.skipif("sys.version_info >= (3,0)")),
])
def test_increment(n, expected):
    assert n + 1 == expected    
```

Surely it is more verbose, but I'm sure it is more natural to read and understand what's going on IMHO, specially for pytest beginners.

One thing that comes to mind is that the current `ids` parameter when given as a function: I understand most people use a short lambda for that, and since now the ids are passed to each `pytest.param` one has to define the lambda before the parametrization and reuse it.

```python
fmt_date = lambda v: v.strftime('%Y/%M/%d')
@pytest.mark.parametrize('date', [
    pytest.param(date(2016, 11, 10), id=fmt_date),
    pytest.param(date(2016, 15, 10), id=fmt_date),
    pytest.param(date(2016, 21, 10), id=fmt_date),
])
```

But that is a minor inconvenience IMHO.

Another issue is that we probably don't want to let users mix styles... if one of the `values` is a `pytest.param` instance, IMO must check that they all are and the `ids` parameter to `pytest.mark.parametrize` is not given. Mixing both (at least initially) seems like to bring confusion to the users.

What are your thoughts?

**Comment by The-Compiler** [comment]
IMHO, combining `pytest.param` and `ids` should work - I don't see a reason to disallow something like:

```python
@pytest.mark.parametrize('date', [
    pytest.param(date(2016, 11, 10), marks=pytest.mark.xfail),
    date(2016, 15, 10),
    date(2016, 21, 10),
], ids=lambda v: v.strftime('%Y/%M/%d'))
```

When `ids` is given and one of the `pytest.param` instances has `id=` used, then probably a failure makes sense.

Other than that, that sounds like a great proposal!

**Comment by RonnyPfannschmidt** [comment]
this is a entirely different concept, but it can be done relatively quickly, i#ll try to finish before the weekend

@nicoddemus did lay out things fabulously

**Comment by nicoddemus** [comment]
> IMHO, combining pytest.param and ids should work

I thought about disallowing it at initially because:

1. It is easier to allow something later if we are OK with it rather than the other way around.
2. It might be weird to allow `pytest.param` with `id=` plus the general `ids=` parameter might be a little strange:

```python
@pytest.mark.parametrize('date', [
    pytest.param(date(2016, 1, 1), id='new years'),
    date(2016, 15, 10),
    date(2016, 21, 10),
], ids=lambda v: v.strftime('%Y/%M/%d'))
```

But after looking at the example itself I think it is pretty obvious that the param-specific `id=` takes precedence. 😉 

Also, as I understand it the idea is to eventually deprecate the current way to mark parametrized values and eventually remove them, so allowing users to mix the two styles is important to make it easy to adopt the new style because they only have to change the parameters that are marked to the new style, instead of all of them:

```python
@pytest.mark.parametrize('date', [
    pytest.mark.xfail()(date(2016, 11, 10)),
    date(2016, 15, 10),
    date(2016, 21, 10),
])
```

To convert it when mixing the styles is allowed:

```python
@pytest.mark.parametrize('date', [
    pytest.param(date(2016, 11, 10), marks=pytest.mark.xfail),
    date(2016, 15, 10),
    date(2016, 21, 10),
])
```

When it is not:

```python
@pytest.mark.parametrize('date', [
    pytest.param(date(2016, 11, 10), marks=pytest.mark.xfail),
    pytest.param(date(2016, 15, 10)),
    pytest.param(date(2016, 21, 10)),
])
```

Depending on the values, a find replace might be tricky as your example with lambdas suggest.

So in summary, I agree with you that we should allow mixing `pytest.param` with the current style. 😁

**Comment by nicoddemus** [comment]
> this is a entirely different concept, but it can be done relatively quickly, i#ll try to finish before the weekend

Sounds good @RonnyPfannschmidt, but perhaps you should hold a bit to gather some more feedback? Particularly @flub which contributed a lot to the discussion already, but would love to hear from @hpk42 and @pfctdayelise as well.

**Comment by hpk42** [comment]
I guess the ``pytest.param()`` or ``pytest.Parameter()`` helper function/Class to cleanly create values with meta-information and allowing a mix until 4.0 makes some sense. But I'd really like to see full documentation on this based on what @nicoddemus already posted.  FWIW I guess it's not going to be a change that breaks too many people's code as marking parametrization values is not done that often.  So warning with 3.1 and erroring out with 4.0 is probably ok.

**Comment by The-Compiler** [comment]
> as marking parametrization values is not done that often

Hmm, I'm not so sure about that - if you use parametrization a lot (which I'd expect many testsuites to do), it's quite natural to want to xfail a single line there, no?

> and allowing a mix until 4.0 makes some sense

Are you talking about marking parameters here, or also about ids? For ids, I think we should still allow mixing `ids=` with `pytest.param()`.

**Comment by RonnyPfannschmidt** [comment]
well, we wouldn't remove the old way, just complain about it for a while

**Comment by hpk42** [comment]
@The-Compiler yes, mixing with ``ids=`` makes sense to me especially since it can be a function as noted earlier.  Basically the only thing that gets deprecated/removed at some later point is the usage of ``pytest.mark.*`` for parametrization values. In any case, i'd really like to see full documentation as part of this PR.

**Comment by nicoddemus** [comment]
@hpk42: 
> Basically the only thing that gets deprecated/removed at some later point is the usage of pytest.mark.* 

I agree, that was my understanding as well.

**Comment by flub** [comment]
Hi all,

I think @nicoddemus' write up looks like a pretty good fleshed out version.  It also provides a pretty good backwards compatibility path so I would agree with the second version where you can completely mix and match, it is more readable then I'd first expect.

Naming is tricky obviously.  Technically you're providing some data to create a TestItem.  But you're providing more then just a parameter, it includes marks as well and who knows what else in the future.  I could possibly describe it as a "partial item" but the parallel with functools.partial is a bit funky since it's the wrong way round.  So after all that musing the only practical thing I'd suggest is to at least use the plural since @nicoddemus' suggestion creates all prameters in one such object.  But maybe the musings inspire someone else ;-)

    @pytest.mark.parametrize(['a', 'b'], [
            pytest.params(1, 2, marks=[pytest.mark.foo, pytest.mark.bar]),
    ])
    def test_foo(self, a, b):
        pass

I wrote the above since none of the earlier examples showed multiple marks on one parametrised item, so just making sure this is how it would look.


Lastly I'd like to thank everyone participating.  I think it is great that we can work on a seemingly glacial pace but still make progress, it is nice to be able to look anew at something.  So thanks for the patience with developing the ideas.

**Comment by The-Compiler** [comment]
> So after all that musing the only practical thing I'd suggest is to at least use the plural since @nicoddemus' suggestion creates all prameters in one such object.

Huh - I don't think it does?

**Comment by flub** [comment]
Well, terminology is a bit tricky I guess, and I did get it wrong.  A parameter is the variable name in the signature while an argument is the object you're passing into it.  Now if you are parametrising (ahem, name conflict) a test function then you're supplying all the arguments at once for each parametrisation.

So I guess the original `pytest.param` can also be read as providing one parametrisation set for the test function.  But it can provide multiple arguments to the test function.

**Comment by nicoddemus** [comment]
We should probably take #349 into account as well

**Comment by RonnyPfannschmidt** [comment]
i am severely opposed, by then we have a multi-mix of many components

the value that is supposed to represent a fixture should be absolutely distinct

i am really sick of having things that mix dozens of concepts in a way that inflicts spaghetti in the implementation for lack of any other way to do it

its already pushing it to dump the test id into it

**Comment by nicoddemus** [comment]
@RonnyPfannschmidt I meant for brainstorming only, didn't mean to imply that `pytest.param` should directly support fixtures.

Or do you mean using fixtures with `parametrize` in general?

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus the issue you referred to is a completely unrelated concept that addresses elements of the parameterset, plus its a can of extra issues on top due to potential parametrization of one of those

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus @hpk42 i updated/rebased the pr and would appreciate a second pair of eyes - i'd like to speed up landing this

File: _pytest/mark.py [hunk 1]
```diff
@@ -4,17 +4,61 @@
 from operator import attrgetter
 from .compat import imap
 
+
 def alias(name):
     return property(attrgetter(name), doc='alias for ' + name)
 
 
+class ParameterSet(namedtuple('ParameterSet', 'values, marks, id')):
```
Comment (nicoddemus) [inline _pytest/mark.py]: I think we should add unittests for `ParameterSet` to make sure we cover all corner cases and possible validation errors.

File: _pytest/python.py [hunk 2]
```diff
@@ -778,7 +778,7 @@ def parametrize(self, argnames, argvalues, indirect=False, ids=None,
             If callable, it should take one argument (a single argvalue) and return
             a string or return None. If None, the automatically generated id for that
             argument will be used.
-            If no ids are provided they will be generated automatically from
+            If no ids are provided they will be generated aut+omatically from
```
Comment (nicoddemus) [inline _pytest/python.py]: Typo: `aut+omatically`

File: _pytest/mark.py [hunk 3]
```diff
@@ -4,17 +4,61 @@
...
+    @classmethod
+    def param(cls, *value, **kw):
```
Comment (nicoddemus) [inline _pytest/mark.py]: Nitpick: should be `*values`

File: _pytest/mark.py [hunk 4]
```diff
@@ -4,17 +4,61 @@
...
+            marks = marks,
+        else:
+            assert isinstance(marks, (tuple, list, set))
+
+        id = kw.pop('id', None)
+        assert not kw, kw
+        return cls(value, marks, id)
+
+    @classmethod
+    def extract_from(cls, param, legacy_force_tuple=False):
```
Comment (nicoddemus) [inline _pytest/mark.py]: I didn't quite get the `legacy_force_tuple` parameter here, even after looking at where it's used. Could add a quick-docstring explaining why it's needed?

Comment (nicoddemus) [inline _pytest/mark.py]: Nitpick: the `param` parameter should be `value`

File: _pytest/mark.py [hunk 5]
```diff
@@ -4,17 +4,61 @@
...
+            argval = argval.args[-1]
+        assert not isinstance(argval, ParameterSet)
+        if legacy_force_tuple:
+            argval = argval,
+
+        return cls(argval, marks=newmarks, id=None)
+
+
+    @property
+    def deprecated_arg_dict(self):
```
Comment (nicoddemus) [inline _pytest/mark.py]: Please add a `_` prefix

File: testing/python/metafunc.py [hunk 6]
```diff
@@ -1438,6 +1443,31 @@ def test_limit(limit, myfixture):
...
+                ),
+            ])
+            def test_increment(n, expected):
+                assert n + 1 == expected
+        """.format(strict=strict)
+        testdir.makepyfile(s)
+        reprec = testdir.inline_run()
+        passed, failed = (0, 2) if strict else (2, 0)
+        reprec.assertoutcome(passed=passed, failed=failed)
+
```
Comment (nicoddemus) [inline testing/python/metafunc.py]: We should add a test for the situation where the user uses passes `id` to a `param` and to `parametrize` in the same call.

File: _pytest/mark.py [hunk 7]
```diff
@@ -4,17 +4,61 @@
...
 
+class ParameterSet(namedtuple('ParameterSet', 'values, marks, id')):
+    @classmethod
+    def param(cls, *value, **kw):
+        marks = kw.pop('marks', ())
+        
+        if isinstance(marks, MarkDecorator):
+            marks = marks,
+        else:
+            assert isinstance(marks, (tuple, list, set))
```
Comment (RonnyPfannschmidt) [inline _pytest/mark.py]: we have no sane mechanism for that point in time, as such i'd just leave the exception until we do get a way to sensibly report such errors

File: _pytest/mark.py [hunk 8]
```diff
@@ -4,17 +4,61 @@
...
+
+        id = kw.pop('id', None)
+        assert not kw, kw
```
Comment (RonnyPfannschmidt) [inline _pytest/mark.py]: while taking a look again, it may be sensible to trigger a pytest warning there

File: _pytest/mark.py (hunk 4 above)
Comment (RonnyPfannschmidt) [inline _pytest/mark.py]: will do, it may need a better name, its for the special case of single name parametrizations

File: _pytest/mark.py (hunk 8 above)
Comment (nicoddemus) [inline _pytest/mark.py]: Hmm I think it should break immediately. From my POV having `**kwargs` here is just a way to emulate the keyword-only syntax of Python 3, so it should bail if the user passes a keyword parameter we don't know about.

File: _pytest/mark.py (hunk 7 above)
Comment (nicoddemus) [inline _pytest/mark.py]: `UsageError` is not a good fit?

File: _pytest/mark.py (hunk 8 above)
Comment (RonnyPfannschmidt) [inline _pytest/mark.py]: then i#ll just raise a Error by applying the kwargs to a function

File: _pytest/mark.py (hunk 7 above)
Comment (RonnyPfannschmidt) [inline _pytest/mark.py]: nope, its an error that does warrant to just stop collection alltogether

Comment (nicoddemus) [inline _pytest/mark.py]: Unless I'm mistaken it seems that's what [UsageError does](https://github.com/pytest-dev/pytest/blob/master/_pytest/config.py#L60), no?

Comment (RonnyPfannschmidt) [inline _pytest/mark.py]: whops, i forgot a "not" there

Comment (nicoddemus) [inline _pytest/mark.py]: Hmm I disagree... if I type `mark=` instead of `marks=`, why would I not want a glaring and loud error at that? Do you have an example of where just emitting a warning makes sense?

Comment (RonnyPfannschmidt) [inline _pytest/mark.py]: i want a collelction error, not a direct stop of the process

Comment (nicoddemus) [inline _pytest/mark.py]: Oh OK, I agree. 

Now I see, I got confused by your comment on the line below:

> while taking a look again, it may be sensible to trigger a pytest warning there

and thought you meant a warning here as well. My bad.

Having collection error at this point is definitely the proper approach. 👍

File: _pytest/mark.py (hunk 5 above)
Comment (RonnyPfannschmidt) [inline _pytest/mark.py]: wont do that as its used in open code and just bad data

File: _pytest/mark.py (hunk 4 above)
Comment (RonnyPfannschmidt) [inline _pytest/mark.py]: extract_from operates on a legacy parameterset, so the name value is certainly wrong

File: doc/en/parametrize.rst [hunk 9]
```diff
@@ -94,21 +94,37 @@ for example with the builtin ``mark.xfail``::
     @pytest.mark.parametrize("test_input,expected", [
         ("3+5", 8),
         ("2+4", 6),
-        pytest.mark.xfail(("6*9", 42)),
+        pytest.param("6*9", 42,
```
Comment (nicoddemus) [inline doc/en/parametrize.rst]: Do you plan to tackle this @RonnyPfannschmidt?

File: doc/en/parametrize.rst [hunk 10]
```diff
@@ -94,21 +94,37 @@ for example with the builtin ``mark.xfail``::
...
+
+        import pytest
+        @pytest.mark.parametrize("test_input,expected", [
+            ("3+5", 8),
+            ("2+4", 6),
+            pytest.mark.xfail(("6*9", 42),),
+        ])
+        def test_eval(test_input, expected):
+            assert eval(test_input) == expected
+
```
Comment (nicoddemus) [inline doc/en/parametrize.rst]: Do you plan to tackle this @RonnyPfannschmidt?

Comment (nicoddemus) [inline doc/en/parametrize.rst]: Just noticed you did already, nevermind my previous comment

File: _pytest/mark.py (hunk 8 above)
Comment (nicoddemus) [inline _pytest/mark.py]: Think this is worthwhile doing?

**Comment by nicoddemus** [comment]
Guys, anybody else wants to take a look at this PR before merging? Unless somebody says otherwise, I will merge this tomorrow. 👍

File: _pytest/mark.py [hunk 11]
```diff
@@ -4,17 +4,71 @@
...
+
+    @classmethod
+    def extract_from(cls, parameterset, legacy_force_tuple=False):
+        """
+        :param parameterset:
+            a legacy style parameterset that may or may not be a tuple,
+            and may or may not be wrapped into a mess of mark objects
+
+        :param legacy_force_tuple:
+          enforce tuple wrapping so single arugment tuple values dont get decomposed and break tests
```
Comment (The-Compiler) [inline _pytest/mark.py]: Some nitpicks: 
- arugment -> argument
- dont -> don't
- indent two spaces more for consistency

File: doc/en/parametrize.rst (hunk 9 above)
Comment (The-Compiler) [inline doc/en/parametrize.rst]: I agree, FWIW.

File: doc/en/parametrize.rst [hunk 12]
```diff
@@ -94,21 +94,42 @@ for example with the builtin ``mark.xfail``::
...
+
+  This was an initial hack to support the feature but soon was demonstrated to be incomplete,
+  broken for passing functions or applying multiple marks with the same name but different parameters.
+  The old syntax will be removed in `pytest 4.0`.
```
Comment (The-Compiler) [inline doc/en/parametrize.rst]: Why the backticks?

File: doc/en/parametrize.rst [hunk 13]
```diff
@@ -94,21 +94,42 @@ for example with the builtin ``mark.xfail``::
...
+
+
 Let's run this::
 
     $ pytest
     ======= test session starts ========
-    platform linux -- Python 3.5.2, pytest-3.0.7, py-1.4.32, pluggy-0.4.0
+    platform linux -- Python 3.5.2, pytest-3.0.3, py-1.4.31, pluggy-0.4.0
```
Comment (The-Compiler) [inline doc/en/parametrize.rst]: Update your pytest! :wink:

File: testing/python/metafunc.py [hunk 14]
```diff
@@ -1438,6 +1450,31 @@ def test_limit(limit, myfixture):
...
+
     def test_pytest_make_parametrize_id(self, testdir):
         testdir.makeconftest("""
             def pytest_make_parametrize_id(config, val):
```
Comment (The-Compiler) [inline testing/python/metafunc.py]: Do we still have tests testing the legacy syntax/compatibility?

Comment (The-Compiler) [inline testing/python/metafunc.py]: Do we still have tests testing the legacy syntax/compatibility?

File: _pytest/mark.py [hunk 15]
```diff
@@ -4,17 +4,71 @@
...
 
     """Error in use of a pytest marker/attribute."""
 
 
+
 def pytest_namespace():
-    return {'mark': MarkGenerator()}
+    return {
+        'mark': MarkGenerator(),
+        'param': ParameterSet.param,
```
Comment (RonnyPfannschmidt) [inline _pytest/mark.py]: well, i deal the conflicts i fold the conflicts ^^

File: _pytest/mark.py (hunk 1 above)
Comment (RonnyPfannschmidt) [inline _pytest/mark.py]: i added a few, perhaps we should make an issue about more detailed tests

File: testing/python/metafunc.py (hunk 14 above)
Comment (RonnyPfannschmidt) [inline testing/python/metafunc.py]: added

Comment (RonnyPfannschmidt) [inline testing/python/metafunc.py]: added

**Comment by RonnyPfannschmidt** [comment]
i am thinking of folding a few commits, any opinion?

**Comment by nicoddemus** [comment]
>i am thinking of folding a few commits, any opinion?

Sure, go ahead!

File: _pytest/python.py [hunk 16]
```diff
@@ -788,36 +788,35 @@ def parametrize(self, argnames, argvalues, indirect=False, ids=None,
...
-            argvalues = [(NOTSET,) * len(argnames)]
-            # we passed a empty list to parameterize, skip that test
-            #
+            force_tuple = len(argnames) == 1
+        else:
+            force_tuple = False
+        parameters = [
+            ParameterSet.extract_from(x, legacy_force_tuple=force_tuple)
+            for x in argvalues]
+        del argvalues
```
Comment (The-Compiler) [inline _pytest/python.py]: What's the point of this?

Comment (RonnyPfannschmidt) [inline _pytest/python.py]: i used this as refactoring aid, the argnames/argvalues variables where badly named and used
just killing it off the local namespace helped finding errors, and it also prevents using them after they shouldnt be used

**Comment by The-Compiler** [comment]
LGTM now, but still needs a changelog entry.
